### PR TITLE
Custom Views: integrate endpoints into front-end

### LINKED ISSUE #4055: Add infinitescroll to ProjectsTable and remove pagination
See if this is doable before all the stores are done. Should be fine - just provide proper properties and callbacks, we can hook it up later.

Requires #4053

### LINKED ISSUE #4056: Update visual side (CSS styles) of ProjectsTable
- consider updating usage of colors to `kobo-common` in place of local `_colors.scss`

Requires #4053

### LINKED ISSUE #4059: Create a configurable ProjectsViewStore store
- data needed:
    - predefined view ID (ultimately we want this to work with arbitrary-ish urls: `/api/v2/assets/?type=survey` vs `/api/v2/assets/?view_uid=7`
- it can handle different lists of assets (some predefined views available on backend)
- it can communicate with `ProjectsTableFilters`, `ProjectsTableFieldsSelector` and `ProjectsTableSorter` - simply informing what the current setting is and receiving requests to change it to new one
- fetching assets, keeping the list up to date
- look at `myLibraryStore.ts` for inspiration, but remember to use MobX
- it should also work with `SearchBox` component (not priority)
- let's not use `assets/hash`

Requires #4050
Requires #4051
Requires #4052
Requires #4063

### LINKED ISSUE #4060: Create a ProjectsViewTable component
- we might not need an actual component, just glueing things together at route component level
- it utilizes all the existing pieces
- data needed:
    - view ID - a way to identify what assets we want
- a wrapper component that can be used anywhere
- it uses `ProjectsTable` together with `ProjectsTableFilters` and `ProjectsTableFieldsSelector`
- data comes from a configurable `ProjectsViewStore` store
- optionally try to have all the params in URL?

Requires #4050
Requires #4051
Requires #4052
Requires #4053
Requires #4059

### LINKED ISSUE #4061: Create DefinedProjectsViewsStore
uses an endpoint that lists all available views and stores the data on app init

We can introduce a built-in views here (like "My Projects"). The difference would be the endpoint from which the data comes from, e.g. `/api/v2/assets/?type=survey` for "My Projects" vs `/api/v2/assets/?view_uid=7` for "North America".

Requires #4064

### LINKED ISSUE #4062: Create new ProjectsViewsRoute route
- uses `DefinedProjectsViewsStore` for the dropdown
- add a "view selector" dropdown for switching betwen defined views ("My Projects", "North america" etc.)
- it utilizes `ProjectsViewTable` passes different IDs
- we should go with route path `/projects/<view ID>`, and `/projects` should redirect to first view (the default My Projects one)

Requires #4053
Requires #4060
Requires #4061

### PR DESCRIPTION
**Pull request by magicznyleszek**
## Description

Hooks up all the necessary parts of Project Views feature to back-end endpoints. Also updates all the missing styles.

## Notes for reviewer

Includes back-end code from `4081-regional-views` branch and few other back-end branches (Olivier merged them here, don't know which are these :P)

This huge mock file with assets is removed (thus the size of removed lines).

~~Please don't merge it after reviewing, the back-end code will need to be reviewed too (probably).~~ All back-end code is already merged into `feature/regional_projects`, so this PR can be merged after being reviewed.

~~Built atop `cleanup-projects-table` branch (PR #4167). Please change base branch to `feature/regional_projects` after that PR is merged.~~

## Related issues

Fixes #4059
Fixes #4060 
Fixes #4061 
Fixes #4062 
Fixes #4055 
Fixes #4056

### CODE REVIEW

**Comment by noliveleger** [comment]
@magicznyleszek back-end code is already reviewed and merged in `feature/regional_projects`

File: jsapp/js/projects/projectsTable/projectsTable.tsx [hunk 1]
```diff
@@ -2,90 +2,93 @@ import React from 'react';
...
+  OrderDirection,
+} from 'js/projects/projectViews/constants';
 import ProjectsTableHeader from './projectsTableHeader';
-import type {AssetResponse} from 'js/dataInterface';
+import type {AssetResponse, ProjectViewAsset} from 'js/dataInterface';
 import styles from './projectsTable.module.scss';
 import rowStyles from './projectsTableRow.module.scss';
 import classNames from 'classnames';
 
+const SCROLL_PARENT_ID = 'projects-table-is-using-infinite_scroll-succesfully';
```
Comment (p2edwards) [inline jsapp/js/projects/projectsTable/projectsTable.tsx:15]:
```suggestion
const SCROLL_PARENT_ID = 'projects-table-is-using-infinite_scroll-successfully';
```

**Review by p2edwards — CHANGES_REQUESTED** [review]

File: jsapp/scss/_colors.scss [hunk 2]
```diff
@@ -1,7 +1,19 @@
-// This is our current pallete of colors.
-//
-// It should be the only source of colors in styles (i.e. no one-shot colors,
-// all `color`, `background-color` etc. should use variables defined here).
+/* DEPRECAED
```
[the remark this answers is not in the document]
Comment (p2edwards) [reply jsapp/scss/_colors.scss:1]: edit: this file is removed entirely in [this PR](https://github.com/kobotoolbox/kpi/pull/4204) anyway

File: jsapp/js/assetUtils.ts [hunk 3]
```diff
@@ -91,9 +92,19 @@ export function getLanguageIndex(asset: AssetResponse, langString: string) {
...
     return asset?.summary?.languages?.join(', ');
+  } else if (
+    asset &&
+    'languages' in asset &&
+    asset.languages.length >= 1
+  ) {
+    return asset.languages.join(', ');
   } else {
     return '-';
   }
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/assetUtils.ts]: Fixed :)

File: jsapp/js/assetUtils.ts [hunk 4]
```diff
@@ -169,7 +180,11 @@ export function getAssetDisplayName(asset: AssetResponse): DisplayNameObj {
   if (asset.name) {
     output.original = asset.name;
   }
-  if (asset?.summary?.labels && asset.summary.labels.length > 0) {
+  if (
+    'summary' in asset &&
+    asset?.summary?.labels &&
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/assetUtils.ts]: Fixed :)

File: jsapp/js/components/formSummary.es6 [hunk 5]
```diff
@@ -87,35 +87,32 @@ class FormSummary extends React.Component {
...
 
       const q2 = `query={"_submission_time": {"$gte":"${lastWeekStart.toISOString()}"}}&fields=["_id"]`;
       dataInterface.getSubmissionsQuery(assetid, q2).done((d) => {
         if (subsCurrentPeriod > 0) {
           let subsPerDay;
-          if (this.state.chartPeriod == 'week')
-            subsPerDay = [0,0,0,0,0,0,0];
-          else
-            subsPerDay = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0];
+          if (this.state.chartPeriod === 'week') {subsPerDay = [0, 0, 0, 0, 0, 0, 0];} else {subsPerDay = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];}
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/formSummary.es6]: Good idea

File: jsapp/js/projects/projectViews/constants.ts [hunk 6]
```diff
@@ -1,85 +1,146 @@
...
+ * EXCEPTION: Dates are special pieces of data and they can be filtered in
+ * a meaningful way by using these queries.
+ */
+export const DATE_FILTER_QUERIES = {
+  greaterThan: '<field>__gt:<YYYY-MM-DD>',
+  greaterOrEqualThan: '<field>__gte:<YYYY-MM-DD>',
+  lessThan: '<field>__lt:<YYYY-MM-DD>',
+  lessOrEqualThan: '<field>__lte:<YYYY-MM-DD>',
+  partOf: '<field>__regex:<YYYY-MM>',
 };
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/projects/projectViews/constants.ts]: I left these WIP constants to not lose these queries. But I made a GH issue with this list included, so I removed it now :)

[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/projects/projectViews/constants.ts]: https://github.com/kobotoolbox/kpi/issues/4212

File: jsapp/js/projects/projectViews/projectsFilterEditor.tsx [hunk 7]
```diff
@@ -36,8 +36,12 @@ export default function ProjectsFilterEditor(props: ProjectsFilterEditorProps) {
...
     props.onFilterChange({
       fieldName: fieldValue,
-      condition: props.filter.condition,
-      value: props.filter.value,
+      // Switching field causes the condition to be dropped - this ensures
+      // we don't end up with unsupported field x condition pair :heart:.
+      condition: undefined,
+      // We also drop the value (if any) as it doesn't make sense to keep it
+      // without condition :shrug:.
+      value: undefined,
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/projects/projectViews/projectsFilterEditor.tsx:44]: I made a GH issue for this: https://github.com/kobotoolbox/kpi/issues/4214

File: jsapp/js/projects/customViewRoute.tsx [hunk 8]
```diff
@@ -1,60 +1,88 @@
...
+  };
 
   const exportAllData = () => {
-    notify.warning(t("Export is being generated, you will receive an email when it's done"));
-    // TODO make the call :)
-    console.log('call backend to initiate downloading data to email');
-  }
+    const foundView = projectViews.getView(viewUid);
+    if (foundView) {
+      $.ajax({
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/projects/customViewRoute.tsx]: 👍

**Review by p2edwards — APPROVED** [review]

File: jsapp/js/dataInterface.ts [hunk 9]
```diff
@@ -549,6 +556,37 @@ export interface AssetResponse extends AssetRequestObject {
...
+  owner__username: string;
+  owner__email: string;
+  /** Full name */
+  owner__name: string;
+  owner__organization: string;
+  uid: string;
+  kind: string;
+  name: string;
+  settings: AssetSettings;
+  languages: Array<string | null>;
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/dataInterface.ts:575]: Languages are never `null`, if no languages are defined, we get empty array. It is possible to get `null` for some old code reasons I think. I'm pretty sure this list is the same as `asset.content.translations`, which is also `Array<string|null>`.

**Pull request merged**